### PR TITLE
*: extract 'RegionCache', add tests

### PR DESCRIPTION
**Pull request by disksing**

### CODE REVIEW

**Comment by disksing** [comment]
PTAL @morefreeze @siddontang @ngaut @BusyJay

File: kv.go [hunk 1]
```diff
@@ -190,7 +154,6 @@ func (s *tikvStore) Close() error {
 	defer mc.mu.Unlock()
 
 	delete(mc.cache, s.uuid)
-	s.regionCache = make(map[uint64]*raftRegion)
 
 	var lastErr error
 	for k, client := range s.clients {
```
Comment (ngaut) [inline kv.go]: s/k/addr

File: region_cache.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+	return nil
+}
+
+func (c *RegionCache) loadRegion(key []byte) (*Region, error) {
+	meta, err := c.pdClient.GetRegion(key)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(meta.Peers) == 0 {
+		return nil, errors.New("recieve Region with no peer")
```
Comment (ngaut) [inline region_cache.go]: s/recieve/receive

File: region_cache.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	if peer != nil {
+		store, err = c.pdClient.GetStore(peer.GetStoreId())
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	delete(c.regions, regionID)
+
+	if peer == nil || err != nil {
+		// Cannot find peer in cached region, or error occurs when loading
```
Comment (ngaut) [inline region_cache.go]: s/Cannot find peer in cached region/Can't find the peer in cache.

**Comment by ngaut** [comment]
Please add split and merge tests. Rest LGTM.

File: region_cache_test.go [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+		Id:      proto.Uint64(id),
+		Address: proto.String(addr),
+	}
+}
+
+func (c *mockPDClient) removeStore(id uint64) {
+	delete(c.stores, id)
+}
+
+func (c *mockPDClient) setRegion(id uint64, startKey, endKey []byte, peers ...uint64) {
```
Comment (morefreeze) [inline region_cache_test.go]: `peers ...` seems no need, it can't support for-loop to generate `peers`, replace by array.

**Comment by morefreeze** [comment]
rest LGTM

**Comment by morefreeze** [comment]
LGTM

**Pull request merged**